pkg/types/contract_reader_cache.go

```diff
@@ -0,0 +1,441 @@
+
+	return nil
+}
+
+func (c *CachedContractReader) fallThroughAndCache(ctx context.Context, key cacheKey, returnVal any) error {
+	data := c.mustGetData(key)
+
+	data.SetLastUpdate(time.Now())
+
+	retBaseType := reflect.Indirect(reflect.New(reflect.Indirect(reflect.ValueOf(returnVal)).Type()))
+	savedVal := reflect.New(retBaseType.Type())
+
+	reflect.Indirect(savedVal).Set(retBaseType)
+
+	cachedValue := savedVal.Interface()
+
+	if err := c.fallThrough(ctx, key.readIdentifier, key.confidence, key.params, cachedValue); err != nil {
+		return err
+	}
+
+	reflect.Indirect(reflect.ValueOf(returnVal)).Set(reflect.Indirect(reflect.ValueOf(cachedValue)))
+	data.SetValue(cachedValue)
+
+	return nil
+}
+
+func (c *CachedContractReader) fallThrough(
+	ctx context.Context,
+	readIdentifier string,
+	confidenceLevel primitives.ConfidenceLevel,
+	params, returnVal any,
+) error {
+	return c.base.GetLatestValue(ctx, readIdentifier, confidenceLevel, params, returnVal)
+}
+
+func (c *CachedContractReader) startSubproc(fn func(), frequency time.Duration) {
+	c.subprocs = append(c.subprocs, make(chan struct{}))
+
+	go func(chClose <-chan struct{}, fn func(), freq time.Duration) {
+		for {
+			timer := time.NewTimer(utils.WithJitter(freq))
+
+			select {
+			case <-timer.C:
+				timer.Stop()
+				fn()
+			case <-chClose:
+				timer.Stop()
+
+				return
+			}
+		}
+	}(c.subprocs[len(c.subprocs)-1], fn, frequency)
+}
+
+func (c *CachedContractReader) clean() {
+	forRemoval := make([]string, 0)
+
+	c.dataLock.RLock()
+
+	for keyStr, data := range c.data {
+		key, keyExists := c.keyLookup[keyStr]
+		if keyExists {
+			if key.staleness == 0 {
+				continue
+			}
+
+			if time.Since(data.LastUpdate()) > key.staleness {
+				forRemoval = append(forRemoval, keyStr)
+			}
+		}
+	}
+
+	c.dataLock.RUnlock()
+
+	c.dataLock.Lock()
+
+	for _, key := range forRemoval {
+		delete(c.data, key)
+		delete(c.keyLookup, key)
+	}
+
+	c.dataLock.Unlock()
+}
+
+func (c *CachedContractReader) pollAndCache() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), pollInterval)
+
+	for _, key := range c.polling {
+		if ctx.Err() != nil {
+			break
+		}
+
+		config, ok := c.getStrategy(key.readIdentifier)
+		if !ok {
+			continue
+		}
+
+		pollConfig := config.(PollAndCache)
+		data := c.mustGetData(key)
+
+		if time.Since(data.LastUpdate()) < pollConfig.Interval {
+			continue
+		}
+
+		_ = c.fallThroughAndCache(ctx, key, pollConfig.Value)
+	}
+
+	cancel()
+}
+
+func (c *CachedContractReader) getStrategy(readIdentifier string) (any, bool) {
+	c.strategyLock.RLock()
+	defer c.strategyLock.RUnlock()
+
+	strategy, ok := c.strategies[readIdentifier]
+
+	return strategy, ok
+}
+
+func (c *CachedContractReader) setStrategy(readIdentifier string, strategy any) {
+	c.strategyLock.Lock()
+	defer c.strategyLock.Unlock()
+
+	c.strategies[readIdentifier] = strategy
+}
+
+func (c *CachedContractReader) mustGetData(key cacheKey) *cachedValue {
```

Comment:
I think `must` will mislead readers in to thinking this "must not error". How about something like `getOrCreate`?